File: 2022/12/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";

// Loads the map
async function load_map(filename) {
	let grid = [];
	let start = {};
	let end = [];
	const stream = fs.createReadStream(filename);
	const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
	reader.on("line", line => {
		grid.push(Array.from(line));
	});
	await new Promise(res => reader.once("close", res));

	for (let y = 0; y < grid.length; y++) {
		for (let x = 0; x < grid[0].length; x++) {
			if (grid[y][x] === "S") { start = { x, y }; grid[y][x] = 'a'; }
			if (grid[y][x] === "E") { end = { x, y }; grid[y][x] = 'z'; }
		}
	}

	let map = { grid, start, end };
	return map;
}

// --- Read and parse the input file ---
let map = await load_map(INPUT_FILE);

// --- Search the smallest path ---
map.findPath = function() {
	const h = this.grid.length;
	const w = this.grid[0].length;
	let a = [];
	for (let y = 0; y < h; y++) {
		let b = [];
		for (let x = 0; x < w; x++) {
			b.push(0);
		}
		a.push(b);
	}
	let x = this.start.x;
	let y = this.start.y;
	a[y][x] = 1;
	let list = [{ x, y }];
	while (list.length > 0) {
		const p = list.splice(0, 1)[0];
		x = p.x;
		y = p.y;
		if (x > 0     && a[y][x - 1] === 0 && this.grid[y][x - 1].charCodeAt(0) <= this.grid[y][x].charCodeAt(0) + 1) { a[y][x - 1] = a[y][x] + 1; list.push({ x: x - 1, y }); }
		if (y > 0     && a[y - 1][x] === 0 && this.grid[y - 1][x].charCodeAt(0) <= this.grid[y][x].charCodeAt(0) + 1) { a[y - 1][x] = a[y][x] + 1; list.push({ x, y: y - 1 }); }
		if (x < w - 1 && a[y][x + 1] === 0 && this.grid[y][x + 1].charCodeAt(0) <= this.grid[y][x].charCodeAt(0) + 1) { a[y][x + 1] = a[y][x] + 1; list.push({ x: x + 1, y }); }
		if (y < h - 1 && a[y + 1][x] === 0 && this.grid[y + 1][x].charCodeAt(0) <= this.grid[y][x].charCodeAt(0) + 1) { a[y + 1][x] = a[y][x] + 1; list.push({ x, y: y + 1 }); }
	}

	x = this.end.x;
	y = this.end.y;
	let c = a[y][x];
	list = [];
	while (c-- > 1) {
		if (x > 0 && a[y][x - 1] == c) x--;
		else if (y > 0 && a[y - 1][x] == c) y--;
		else if (x < w - 1 && a[y][x + 1] == c) x++;
		else if (y < h - 1 && a[y + 1][x] == c) y++;
		list.push({ x, y });
	}
	return list;
};

// --- Searches all the paths from the lowest elevation points ---
map.findAllPaths = function() {
		const h = this.grid.length;
		const w = this.grid[0].length;
		let list = [];
		for (let y = 0; y < h; y++) {
			for (let x = 0; x < w; x++) {
				if (this.grid[y][x] == 'a') {
					this.start = { x, y };
					let path = this.findPath();
					if (path.length > 0) list.push(path);
				}
			}
		}
		return list;
};

// --- Puzzle 1 ---
const count = map.findPath().length;
console.log("1. Fewest steps: ", count);

// --- Puzzle 2 ---
const min = Math.min(...map.findAllPaths().map(x => x.length));
console.log("2. Shortest path: ", min);

File: 2022/12/rust/main.rs
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants ---
const INPUT_FILE:&'static str = "../input/input.txt";

// --- Types ---
type Point = (usize, usize);
type Grid = Vec<Vec<char>>;

// --- Structs ---
struct Map {
    grid: Grid,
    start: (usize, usize),
    end: (usize, usize)
}

impl Map {
    fn load_from_file(filename: &str) -> Result<Self, std::io::Error> {
        let file = File::open(filename)?;
        let mut grid: Grid = BufReader::new(file).lines()
            .map(|line| line.unwrap().chars().collect())
            .collect();
        let mut start: Option<Point> = None;
        let mut end: Option<Point> = None;
        for y in 0..grid.len() {
            for x in 0..grid[0].len() {
                if grid[y][x] == 'S' { start = Some((x, y)); grid[y][x] = 'a'; }
                if grid[y][x] == 'E' { end   = Some((x, y)); grid[y][x] = 'z'; }
            }
        }
        if start.is_none() || end.is_none() {
            panic!("Missing Starting or Ending point");
        }
        Ok(Map { grid, start: start.unwrap(), end: end.unwrap() })
    }

    fn find_path(&self) -> Vec<Point> {
        let h = self.grid.len();
        let w = self.grid[0].len();
        let x = self.start.0;
        let y = self.start.1;
        let mut a: Vec<Vec<usize>> = vec![vec![0; w]; h];
        a[y][x] = 1;
        let mut list: Vec<Point> = vec![ self.start ];
        while list.len() > 0 {
            let p = list.remove(0);
            let x = p.0;
            let y = p.1;
            if x > 0     && a[y][x - 1] == 0 && self.grid[y][x - 1] as usize <= self.grid[y][x] as usize + 1 { a[y][x - 1] = a[y][x] + 1; list.push((x - 1, y)); }
            if y > 0     && a[y - 1][x] == 0 && self.grid[y - 1][x] as usize <= self.grid[y][x] as usize + 1 { a[y - 1][x] = a[y][x] + 1; list.push((x, y - 1)); }
            if x < w - 1 && a[y][x + 1] == 0 && self.grid[y][x + 1] as usize <= self.grid[y][x] as usize + 1 { a[y][x + 1] = a[y][x] + 1; list.push((x + 1, y)); }
            if y < h - 1 && a[y + 1][x] == 0 && self.grid[y + 1][x] as usize <= self.grid[y][x] as usize + 1 { a[y + 1][x] = a[y][x] + 1; list.push((x, y + 1)); }
        }

        let mut x = self.end.0;
        let mut y = self.end.1;
        let mut c = a[y][x];
        list.clear();
        while c > 1 {
            if x > 0 && a[y][x - 1] == c { x -= 1; }
            else if y > 0 && a[y - 1][x] == c { y -= 1; }
            else if x < w - 1 && a[y][x + 1] == c { x += 1; }
            else if y < h - 1 && a[y + 1][x] == c { y += 1; }
            list.push((x, y));
            c -= 1;
        }
        list
    }

    fn find_all_paths(&mut self) -> Vec<Vec<Point>> {
        let h = self.grid.len();
        let w = self.grid[0].len();
        let mut list: Vec<Vec<Point>> = vec![];
        for y in 0..h {
            for x in 0..w {
                if self.grid[y][x] == 'a' {
                    self.start = (x, y);
                    let path = self.find_path();
                    if path.len() > 0 { list.push(path); }
                }
            }
        }
        list
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // --- Parse the input file ---
    let mut map = Map::load_from_file(INPUT_FILE)?;

    // --- Puzzle 1 ---
    let count = map.find_path().len();
    println!("1. Fewest steps: {}", count);

    // --- Puzzle 2 ---
    let min = map.find_all_paths().iter().map(|x| x.len()).min().unwrap();
    println!("2. Shortest path: {}", min);

    Ok(())
}
